phabricatoremails/template.py

```diff
@@ -0,0 +1,237 @@
+    return reviewer.is_actionable and reviewer.status == ReviewerStatus.UNREVIEWED
+
+
+def is_reviewer_needed(reviewer: Reviewer):
+    # This specific reviewer needs to accept this revision before it can be landed
+    return reviewer.is_actionable and reviewer.status in [
+        ReviewerStatus.REQUESTED_CHANGES,
+        ReviewerStatus.BLOCKING,
+    ]
+
+
+def reviewer_actionable_class(reviewer: Reviewer):
+    return "actionable" if reviewer.is_actionable else "non-actionable"
+
+
+def date(utc_value: datetime, receiver_timezone: timezone):
+    return utc_value.astimezone(receiver_timezone).strftime("%b %-d %-I:%M%p:")
+
+
+def comment_summary(body):
+    if not body.main_comment and not body.inline_comments:
+        return ""
+    elif (body.main_comment and not body.inline_comments) or (
+        not body.main_comment and len(body.inline_comments) == 1
+    ):
+        return " and submitted a comment"
+    else:
+        return " and submitted comments"
+
+
+def secure_comment_summary(body):
+    if body.comment_count == 0:
+        return ""
+    elif body.comment_count == 1:
+        return " and submitted a comment"
+    else:
+        return " and submitted comments"
+
+
+def diff_class(diff_type: DiffLineType):
+    if diff_type == DiffLineType.ADDED:
+        return "added-line"
+    elif diff_type == DiffLineType.REMOVED:
+        return "removed-line"
+    elif diff_type == DiffLineType.NO_CHANGE:
+        return "no-change-line"
+
+
+def reviewer_status_icon(status: ReviewerStatus):
+    if status == ReviewerStatus.ACCEPTED:
+        return "&#x2714;"
+    elif status == ReviewerStatus.REQUESTED_CHANGES:
+        return "&#x2716;"
+    else:
+        return ""
+
+
+def existence_change_class(reviewer_change: ExistenceChange):
+    if reviewer_change == ExistenceChange.NO_CHANGE:
+        return ""
+    elif reviewer_change == ExistenceChange.ADDED:
+        return "added"
+    elif reviewer_change == ExistenceChange.REMOVED:
+        return "removed"
+
+
+def existence_change_label(reviewer_change: ExistenceChange):
+    if reviewer_change == ExistenceChange.NO_CHANGE:
+        return ""
+    elif reviewer_change == ExistenceChange.ADDED:
+        return "added"
+    elif reviewer_change == ExistenceChange.REMOVED:
+        return "removed"
+
+
+def file_change(change: AffectedFileChange):
+    if change == AffectedFileChange.MODIFIED:
+        return "modified"
+    elif change == AffectedFileChange.ADDED:
+        return "added"
+    elif change == AffectedFileChange.REMOVED:
+        return "removed"
+
+
+def diff_symbol(diff_type: DiffLineType):
+    if diff_type == DiffLineType.ADDED:
+        return "+"
+    elif diff_type == DiffLineType.REMOVED:
+        return "-"
+    elif diff_type == DiffLineType.NO_CHANGE:
+        return ""
+
+
+def text_comment(comment: str):
+    # Email width is 80, I use two characters for "> ". So, each line gets 78 characters
```

Comment:
[Avoid](https://developers.google.com/style/pronouns#personal-pronouns) using "I" in comments.

`78` should be stored in a variable somewhere with a descriptive name (e.g. `MAX_EMAIL_CHARS_PER_LINE`.

Reply:
Why is the separate variable needed?
Sure, the `MAX_EMAIL_CHARS_PER_LINE` is self-descripting, but that's also why I've added the comment there to provide the much-needed context

Reply:
This seems like an important constant value to point out to a developer perusing the code, and would probably even consider putting it at the top of the file along with other constants in this module.

I generally avoid hard coding any value that has an impact on the operation of the application. Even if the value may not need to be changed now, it may need to be changed in the future, or perhaps never. But defining this in its own variable will be a small step away from hard coding.

Reply:
> This seems like an important constant value to point out to a developer perusing the code, and would probably even consider putting it at the top of the file along with other constants in this module.

Putting it at the top of the file would remove it from the context from where it's used.

I don't understand why you'd want to group up all the constants together. You should want to set up your code into logical groups . Taking the advice further, should I move _all_ constants in `phabricator-emails` into a single `constants.py` file? Should I group all functions in `functions.py`?

To make an unserious metaphor, this is like having a few sandwiches, then re-ordering them such that all the bread is together, cheese together, mayonnaise together, and so on.

> I generally avoid hard coding any value that has an impact on the operation of the application. Even if the value may not need to be changed now, it may need to be changed in the future, or perhaps never.

Once a value is used by enough different components, I agree that it should be exposed so that it can be re-used.

Reply:
> Taking the advice further, should I move all constants in phabricator-emails into a single constants.py file? Should I group all functions in functions.py?

I assume this is a rhetorical question.

> Putting it at the top of the file would remove it from the context from where it's used.

The module context itself should be sufficient to keep sense of the logical grouping. I'm not saying it's a requirement to move them to the top of the file, but it is a common convention.

However, putting the value in a named variable is less about convention and more about good practice. At least move these values into a variable.

Reply:
> I assume this is a rhetorical question.

Heh, was missing the obligatory :wink: in there :)

> However, putting the value in a named variable is less about convention and more about good practice. At least move these values into a variable.

I think that this conversation can be safely grouped into [this thread](https://github.com/mozilla-conduit/phabricator-emails/pull/1#discussion_r404936558)